circuit/src/merkle_tree/native.rs

```diff
@@ -0,0 +1,158 @@
+use std::marker::PhantomData;
+
+use halo2wrong::halo2::arithmetic::FieldExt;
+
+use crate::{params::RoundParams, poseidon::native::sponge::PoseidonSponge};
+
+const WIDTH: usize = 5;
+
+#[derive(Clone, Debug)]
+struct MerkleTree<F: FieldExt, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	_params: PhantomData<P>,
+	root: Option<F>,
+	indexes: Vec<Option<F>>,
+	height: usize,
+	first_leaf_index: usize,
+}
+
+impl<F: FieldExt, P> MerkleTree<F, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	fn new(height: usize) -> Self {
+		let mut first_leaf_index = 1;
+		let mut nodes = 1;
+
+		for i in 0..height {
+			first_leaf_index *= 2;
+			nodes = 2 * nodes;
+		}
+		first_leaf_index -= 1;
+		nodes *= 2;
+		let indexes = vec![None; nodes - 1];
+		MerkleTree { _params: PhantomData, root: None, indexes, height, first_leaf_index }
+	}
+
+	fn implement_leaves(self: &mut MerkleTree<F, P>, values: Vec<F>) -> &Self {
+		for i in 0..values.len() {
+			self.indexes[self.first_leaf_index + i] = Some(values[i]);
+		}
+		self
+	}
+
+	fn build_tree(self: &mut MerkleTree<F, P>, height: usize) -> &Self {
+		if height > 0 {
+			let mut new_tree = MerkleTree::<F, P>::new(height - 1);
+			let first_leaf_index = (new_tree.first_leaf_index * 2) + 1;
+			let mut hash = Vec::new();
+			for i in 0..first_leaf_index {
+				let mut hasher: PoseidonSponge<F, WIDTH, P> = PoseidonSponge::new();
+				if i % 2 == 0 {
+					hasher.update(&[
+						self.indexes[first_leaf_index + i].unwrap(),
+						self.indexes[first_leaf_index + i + 1].unwrap(),
+					]);
+					hash.push(hasher.squeeze());
+				}
+			}
+			new_tree.implement_leaves(hash.clone());
+			for i in new_tree.first_leaf_index..new_tree.indexes.len() {
+				self.indexes[i] = new_tree.indexes[i];
+			}
+			self.build_tree(new_tree.height);
+		}
+		self.root = self.indexes[0];
+		self
+	}
+
+	fn find_path(self: &mut MerkleTree<F, P>, value: Option<F>, index: Option<usize>) -> Vec<F> {
```

Comment:
1) `self: &mut MerkleTree<F, P>` can be just replaced by `&mut self`. Same goes for other functions

2) I guess you wanted to have the option to find path either by value or index. I think having only by value is sufficient for our use-case